### DISCUSSION TITLE
[Hardware Config] MA12070P Board with Waveshare ESP32 One

### DISCUSSION THREAD

**Discussion by haldi4803** [sle118/squeezelite-esp32#358] [General]
Howdy,

After getting the ESP Audio Dock with [TAS5805M](https://github.com/sle118/squeezelite-esp32/discussions/357) to work i've bought another board with Merus Audio MA12070P onboard.
Because 2x80W is louder than 2x23W Right?.....
Well anyway... the board i wanted was this one:

![Test](https://github.com/sle118/squeezelite-esp32/assets/17104473/89535973-cb75-43cf-80d3-5fa179538a17)

Sadly AliExpress doesn't deliver what one orders but what the seller has in stock...
So i got the 2 Boards Edition.

![image](https://github.com/sle118/squeezelite-esp32/assets/17104473/389ab377-88d4-40df-9d3c-5c5d1221c3a4)

Doesn't matter anyway as neither of the Boards has PS-RAM -.-

So for Trial and Error i've plugged the Board on a Raspberry Pi 4 (Pin compatible! yay) with MoodeAudio installed.
Was kinda annoyed by the 3W Powerdrain while Raspby and Board are running and ordered a [WaveShare ESP32 One](https://www.waveshare.com/wiki/ESP32_One) (which is also Pin Compatible to Raspberry Pi and does have 8MB PS-RAM onboard.) 
![esp32-one-kit-1](https://github.com/sle118/squeezelite-esp32/assets/17104473/7c5fe0f2-8257-4c24-81a6-264c72eb9469)

Flashing SqueezeliteESP32 worked flawless via Web interface.
Now powerdraw seems to be around 0.5W with WiFi connect but nothing set up yet.

Luckily i found the config for the Pins on the AliExpress Page: (The colored one is for WaveShare ESP32 One)
Nothing stopping us now from configuring an I2S Dac.

![Pin_Header](https://github.com/sle118/squeezelite-esp32/assets/17104473/17cdc0e1-2cbb-44ea-a99e-e06c65857edc)
![ESP32-One-Pins](https://github.com/sle118/squeezelite-esp32/assets/17104473/eaceb1ab-b644-4965-a522-f6bb6e33b62f)

According to these we set I2C Ports to GPIO18 for SDA and GPIO23 for SCL.
Yet nothing detected on I2C Scan, because AMP is still offline.
Pin 8 is Enable, so we set GPIO01 to High via set_GPIO in nvConfig. `1=vcc` commit, reboot and we can see the vcc pin in the Status Tab.

According to documentation I2C Adress can be configured by AD0 and AD1.
![image](https://github.com/sle118/squeezelite-esp32/assets/17104473/bb91c838-bf07-4d95-a7a8-1c75fb8fe289)
~~I assume those are CFG2 and CFG3 but doesn't matter as both on Low = 0x20 aka 32~~

Still nothing showing up... so time for Official [Documentation](https://www.infineon.com/dgdl/Infineon-MA12070P-DS-v01_00-EN.pdf?fileId=5546d46264a8de7e0164b761f2f261e4):
Luckily [someone](https://github.com/jorgenkraghjakobsen/merus_bt) already made the MA12070P working on ESP32. That might help as well...

### Start-up procedure
It is recommended to follow the start-up procedure as described below:
1) Make sure the all hardware pins are configured correctly: e.g. BTL, Slave Clock mode.
2) Keep the device in disable and mute: /ENABLE = 1; /MUTE = 0.
3) Bring up 5V VDD supply and PVDD supply (it does not matter if VDD or PVDD comes up first, provided that
the device is held in disable).
4) Wait for VDD and PVDD to be stable.
5) CLK must be present before enabling the amplifier.
6) Enable device: /ENABLE = 0.
7) Program applicable initialization to registers.
8) Unmute device: /MUTE = 1.
9) The device is now in normal operation state.

So we can see that we need to set the GPIO01 aka pin8 to High to start the Device, but then need to put it to Low to access the registers...
Time to test out Philippe's Testbuild with GPIO Control in DAC_Controlset -.-

Edit:
Okay....
Setting GPIO1 to amp in set_GPIO nvs apparently does the trick with booting it up correctly to show in i2C.
![image](https://github.com/sle118/squeezelite-esp32/assets/17104473/be359a9a-7b76-49ea-b75a-db77719e8905)

Using: `1=power:0,3=amp:1` does the trick with powering the Amp. But configuration still seems to be off. as i only get noise :(

According to Documentation there are some i2S configurations that could be made...
![image](https://github.com/sle118/squeezelite-esp32/assets/17104473/c9a5b5a7-13b8-4968-ab99-146bd9b7857b)
![image](https://github.com/sle118/squeezelite-esp32/assets/17104473/630a7a28-3026-4767-9589-a450da8b7465)

Sadly i have no clue to which ones are what ^^

Edit 2:
P.S MSEL0 needs to be high to enter BTL mode instead of PBTL.... 

Edit 3:
going by the Procedure the [other guy](https://github.com/jorgenkraghjakobsen/merus_bt/blob/master/components/MerusAudio/MerusAudio.c#L53) used for his Bluetooth player startup should be:
Set mute to 0
Set enable to 1
i2C communication.
Set enable to 0

Managed to get as far as that as i can read/write to the chip on 0x20 via Web interface...

the read Hardware version is irrelevant i assume. just skip that.

 ma_write_byte(MA120X0_ADDR,1,MA_i2s_format__a,8);          // Set i2s left justified, set audio_proc_enable
 ma_write_byte(MA120X0_ADDR,1,MA_vol_db_master__a,0x28);    // Set vol_db_master

Set Register 53 to 8 .... not sure if that does anything, according to documentation 8 is not a valid input.
Set register 64 to 40 which would just be a volume adjustment.

   res = ma_read_byte(MA120X0_ADDR,1,MA_error__a);
   printf("Errors : 0x%02x\n",res);

Read register 124 aka 7C in my case i always get 0x04 retour. Which would mean PLL Error.

Edit4:
Using the TestBuild 1613 i got a working boot sequence.
set_GPIO: `1=vcc, 3=vcc`
dac_controlset: `{"init":[{"gpio":3,"level":0},{"gpio":1,"level":0},{"reg":53,"val":0},{"reg":54,"val":17},{"gpio":3,"level":1}]}`
Doing that results in register 124 (errors) to be 0x00 so safe.

Only problem remaining is that when i set MSEL0 to High on GPIO5 there is no output at all and Register 117 (aka 0x75) results in 101.
If i leave it at low there is noise in the output :(

Edit5:
I found this post in DiY Audio Forum where Michael Xiao claimed to be the designer of the Board, i wrote him an E-Mail and got a prompt response. Including the Schematics
[SCHEMATIC1 - MA12070[1400].pdf](https://github.com/sle118/squeezelite-esp32/files/13318978/SCHEMATIC1.-.MA12070.1400.pdf)

MSEL1 is fixed to Ground. So MSEL0 needs to be high.
![image](https://github.com/sle118/squeezelite-esp32/assets/17104473/401ce561-994b-4f79-bcd9-312eff5c19e2)

AD0 and AD1 are fixed to Ground so changing i2C adress is not possible.
![image](https://github.com/sle118/squeezelite-esp32/assets/17104473/4582e430-0518-41e3-8c20-eff4202c143e)

BUT I found something else about CFG2 and CFG3... Voltage Control for VDD224 which is the USB-PD Chip.
![image](https://github.com/sle118/squeezelite-esp32/assets/17104473/efeda7f6-766e-4817-8c7b-0fe513a9125d)
Not exactly sure if it matters, but setting CFG2 aka pin13 aka GPIO15 to high now.

**Comment by philippe44** [comment]
If you have noise, it's maybe just the fact that standard must be i2s_format (00) - not left-justified and i2s_framesize must be 32 (10), unless you use the 32 bits version of our software. The default seems to be fine for the others.

**Reply by haldi4803** [reply]
Indeed. After Setting Adress 53 to 0 and adress 54 to 17 i got sound output. Thanks for clarifying that.

But there is still a lot of noise while playing.
i aussume i still have some issues with BTL Configuration.
If i set gpio5 to high for MSEL0 to be high i don't get any output at all... if i set it to low i get noisy output.
i think MSEL1 might be high as well... but i have no clue which gpio that is... maybe i can trace it on the board.

**Comment by sle118** [comment]
For the record, it should be noted that the built-in ESP32 isn't being used since it is a WROOM32 and doesn't have PSRAM. It was very close to being an interesting standalone board.

**Comment by DerPicknicker** [comment]
@haldi4803 … could you link that board on AliExpress or the shop?

edit : is your noise issue fixed?

**Reply by haldi4803** [reply]
Nope.
Never managed to get it working.
But when used with a Raspberry it didnt really provide more output power than the Louder-ESP so i gave up.

here:
https://www.aliexpress.com/item/1005001810845146.html

**Reply by DerPicknicker** [reply]
So you wouldn't recommend it. Good to know. Then I will stay with LOUDER ;-)

**Reply by bench2012** [reply]
ok 3 year late to this game, but finally got it working with the help from DIY Audio #mboxler. The magic is in the clock rates ie. need to add -R -r "96000-96000" to the end of the autoexec and it works!

**Reply by haldi4803** [reply]
> need to add -R -r "96000-96000" to the end of the autoexec and it works!

Wow...
Yeah autoexec1 
I can confirm this!
All noise gone.

Sadly its after 10pm here so I can't try out full volume :(

P.S Great.... Now I got 5 working boards and only 3 Speaker pairs...

**Reply by bench2012** [reply]
Great, just curious, how well does the Waveshare board you are using work? I got a updated version of the ESP32 controller board from the same seller, which use a Wrover, but they really mess up the GPIO (they use the same gpio pin for i2s and the rotary encoder, WTF?!). I rewire it to make both rotary work and added an i2c for OLED. That max out all the GPIO. Would be nice to know if the Waveshare fair better.

**Reply by haldi4803** [reply]
> Great, just curious, how well does the Waveshare board you are using work?

Absolutly Horrible?
It does not Power via 19V Directly, WiFi works but no Sound, sound only Works when connected to USB-TypeC Power.
Soundquality is terrible works until 60% Volume and above it probably doesn't get enough Power and shuts down. But then again at around 50% volume you get horrible clipping, it sounds awful. I'm pretty sure something is still wrong with the config.
About the absolute Volume i can't really say anything as it never worked.

Honestly just go with a Louder-ESP32 module from Tindie from Sonocotta. Those 2x25W are loud enough. Though they also have some power issue on max volume ^^ although more expensive the newer Esparagus might be better.

**Comment by AndreyLapshov** [comment]
Hello, can u provide names and links for 4 boards that u wanted and got?

**Reply by bench2012** [reply]
Got mine from TaoBao, but honestly, I wouldn't bother. The wrover board is a real mass, I mean they provide two rotary encoder onboard, but none of them will work, since they share GPIOs with the i2s bus. I need to rewire them to get them to work. The MA12070P board is nothing special.

Here the link for all it worth: https://item.taobao.com/item.htm?id=643111582592&mi_id=0000EtQ5WI7wjF6mFgO6a22wwgzjq6Xt4kj3tst0Q8264lE&spm=tbpc.boughtlist.suborder_itemtitle.1.26d94c27s5Jx4J

**Comment by bench2012** [comment]
I am powering it through the USB-C PD (the MA12070P board and the Wrover board) and it's fine to max volume. Not very loud, but it could be my speakers, which are a pair of  JAMO 20W 4Ohm and/or my PSU which only supply upto 20V. Not tested the DC input yet. Might try that next, since it suppose to take upto 26V, I got a 24V barrel jack SMP PSU kicking around somewhere.
 
I knock up a few TDA5805M board and will be trying these. Esparagus is nice, but quite expensive.

BTW this is the dac_controlset, hope it helps: { "init": [{"reg":53,"val":136}, {"reg":54,"val":81} , {"reg":64,"val":36}, {"reg":45,"val":52}, {"reg":45,"val":48}] }
